handler/src/main/java/io/netty/handler/ssl/OpenSslSessionContext.java

```diff
@@ -41,30 +40,73 @@
     private final OpenSslKeyMaterialProvider provider;
 
     final ReferenceCountedOpenSslContext context;
+    final OpenSslNullSession nullSession;
+
+    final OpenSslSessionCache sessionCache;
+    private final long mask;
 
     // IMPORTANT: We take the OpenSslContext and not just the long (which points the native instance) to prevent
     //            the GC to collect OpenSslContext as this would also free the pointer and so could result in a
     //            segfault when the user calls any of the methods here that try to pass the pointer down to the native
     //            level.
-    OpenSslSessionContext(ReferenceCountedOpenSslContext context, OpenSslKeyMaterialProvider provider) {
+    OpenSslSessionContext(ReferenceCountedOpenSslContext context, OpenSslKeyMaterialProvider provider, long mask,
+                          OpenSslSessionCache cache) {
         this.context = context;
         this.provider = provider;
+        this.mask = mask;
         stats = new OpenSslSessionStats(context);
+        sessionCache = cache;
+        // If we do not use the KeyManagerFactory we need to set localCertificateChain now.
+        // When we use a KeyManagerFactory it will be set during setKeyMaterial(...).
+        nullSession = new OpenSslNullSession(this, provider == null ? context.keyCertChain : null);
+        SSLContext.setSSLSessionCache(context.ctx, cache);
+    }
+
+    final DefaultOpenSslSession newOpenSslSession(long sslSession, String peerHost,
+                                            int peerPort, String protocol, String cipher,
+                                            OpenSslJavaxX509Certificate[] peerCertificateChain, long creationTime) {
+        return sessionCache.newOpenSslSession(sslSession, this, peerHost, peerPort, protocol, cipher,
+                peerCertificateChain, creationTime);
+    }
+
+    @Override
+    public void setSessionCacheSize(int size) {
+        ObjectUtil.checkPositiveOrZero(size, "size");
+        sessionCache.setSessionCacheSize(size);
```

Comment:
This methods may invoke `freeSessions()` internally. In all other cases you protect the `freeSessions` with write lock. Do we need a lock here as well?

Reply:
`freeSessions()` is defined as `synchronized` is this not good enough ?

Reply:
missed that, 👍